onnxruntime/core/providers/cpu/math/transpose_matmul.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "transpose_matmul.h"
+
+#include "core/framework/op_kernel_context_internal.h"
+#include "core/providers/cpu/math/matmul_helper.h"
+#include "core/util/math.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    TransposeMatMul,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+    TransposeMatMul);
+
+TransposeMatMul::TransposeMatMul(const OpKernelInfo& info)
+    : OpKernel{info} {
+  ORT_ENFORCE(info.GetAttr("transA", &trans_a_attr_).IsOK());
+  ORT_ENFORCE(info.GetAttr("transB", &trans_b_attr_).IsOK());
+}
+
+Status TransposeMatMul::Compute(OpKernelContext* context) const {
+  auto* context_internal = dynamic_cast<OpKernelContextInternal*>(context);
```

Comment:
Is this the only place that uses dynamic_cast?

Reply:
Can you please clarify the question?

---
In reply to: [392564686](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r392564686) [](ancestors = 392564686)

Reply:
Usually we don't use dynamic_cast in our code. But it's fine if you feel it's necessary. Because dynamic_cast need RTTI, we may disable RTTI in the future to reduce binary size further.

Reply:
Per discussion w/ Ed, sometimes we don't get an OpKernelContextInternal, thus needs dynamic_cast. It's rare case. For example - [https://github.com/microsoft/onnxruntime/blob/88c65f8add003f77e1749a5321095e0fcfbfd0e7/onnxruntime/core/optimizer/constant_folding.cc#L71](https://github.com/microsoft/onnxruntime/blob/88c65f8add003f77e1749a5321095e0fcfbfd0e7/onnxruntime/core/optimizer/constant_folding.cc#L71)

---
In reply to: [395776822](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r395776822) [](ancestors = 395776822)